### ISSUE TITLE
Keybase desktop is a huge vulnerability

### ISSUE THREAD

**Issue by jodavaho** [keybase/keybase-issues#3135]
The native desktop application does not require *any* login verification. Upon opening, a malicious user could revoke all verification, change the passphrase for the account, and revoke all gpg keys. This effectively erases a user.

The concept of installing on many devices to ensure the stability of your identity chain basically becomes a huge attack surface without some form of login verification. At minimum, a pin should be unique on each device.

**Comment by zQueal** [comment]
> The concept of installing on many devices to ensure the stability of your identity chain basically becomes a huge attack surface

All of this is assuming someone has access to your PC. Of which, if they have physical access they could basically find a way to do this regardless of any other factors. Chrome saved passwords, bad password conventions, etc.

I wouldn't necessarily call this an attack vector--that and the "safeguard" is as simple as not letting anyone with possible malicious intentions use your user account.

Unless what you're referring to by login verification is 2FA--of which we're in total agreement. Lack of 2FA is the reason I don't use Keybase and just monitor it.

**Comment by jodavaho** [comment]
> All of this is assuming someone has access to your PC. Of which, if they have physical access they could basically find a way to do this regardless of any other factors. Chrome saved passwords, bad password conventions, etc.

Not necessarily, right? I personally don't have sign-in on chrome or firefox. My passwords are all master-key protected, but that's just me. I have exercised the option of doing so. Keybase providing zero *options* is irresponsible. 

My security - conscious applications would do the same. Without some friction, even a small amount, it's too easy. Given the large number of services associated with keybase, it makes sense to give *options* to increase the friction.

I'd suggest, at minimum, a pin for each device:

Keybase desktop app (and mobile app) should be able to configure a non-synchronized pin for each device which is stored locally-only. Even my worst banking apps do this. Even my IM app does this. Losing the pin means re-installing the app, and signing in per the usual *for that device*. 

I'd also like if keybase did not allow sign-ins on new devices without adding the device from another device, CAVEAT: I just don't remember if that's the case. If not *why not?*. That's no more secure than any crap web service with 2FA, and it *already has the functionality to do this*.

**Comment by oconnor663** [comment]
> I'd also like if keybase did not allow sign-ins on new devices without adding the device from another device, CAVEAT: I just don't remember if that's the case.

You're right, it's not possible to sign in on a new device without cooperation from an existing device.

It's also possible to log out on an existing device. To log back in, you'll need your password. There might be an obscure environment variable somewhere that would make your login go away on shutdown; let me know if that would be useful to you.

**Comment by zQueal** [comment]
> However, an app that allows password-less access to an account by default, and has no options to enable passwords, is in my mind a vulnerability by default.

It's not really password-less. You initially need to login--for which you need your password. The real issue you seem to have is the idea that people with physical access to your computer would be able to change or revoke your identity. Which is understandable, really. 

> Keybase providing zero options is irresponsible.

This we can both agree on. Keybase needs the __options__ for those who seek to employ them. Disable automatic login, enable 2FA, and other security measures are an absolute must before anyone can really take Keybase seriously.

2FA was pretty high on the road map a long time ago. But it's been a year or two since I've heard anything about it. IIRC at the time Authy was being considered--which hopefully is no longer the case. But the fact remains a bare minimum of 2FA is necessary.

**Comment by junderw** [comment]
Simple solution: add auto-logout as an option.

Then you need to re-enter your password everytime you want to unlock.

**Comment by rivy** [comment]
+1 for 2FA.

@zQueal , why would Authy be a bad choice?

**Comment by zQueal** [comment]
It's not really smart to do anything that locks you into a particular vendor. Unless they've changed the Authy format that allows you to use alternative software based TOTP applications and not strictly the proprietary Authy one.

On one hand you can use TOTP with the Authy app, but what about anyone who dislikes Authy? Or doesn't want to store their TOTP seed on Authy servers? On the other hand, if you implement TOTP you can use any authenticator application in addition to Authy if someone wanted to go that route.

It's basically having multiple options vs having only one option argument.

**Comment by rivy** [comment]
Ah! I understand. I thought you meant the application itself. I completely agree that a proprietary 2FA would be a bad idea. Thanks for the explanation.

TOTP, and multiple compatible authenticators, would be my choice as well.

Just FYI, the Authy application can now be used for TOTP (I'm using it instead of the Google Authenticator application for TOTP tokens because it has multiple device support and synchronization).

**Comment by jasoncoposky** [comment]
+1 for a basic pin code on the phone app, which is fairly standard for apps which may contain sensitive information, such as dropbox.

**Comment by junderw** [comment]
Basic PIN for all apps, and a setting to set how many minutes until it is set. While PIN-locked, terminal keybase commands are also restricted.

PIN should have a fail limit of 3 or so. After 3 failed attempts, the user is forced to logout. Killing the session and requiring the password to get back in.

Also, fingerprint scan, when activated and available, will take the place of the PIN. If Fingerprint scan fails or device doesn't support, fall back to PIN. (Fingerprint scan fails don't count toward "auto-logout" fail count)

also, PIN should be minimum 4 digits with no maximum. Maybe have a green "Enter" button only light up once 4 digits are entered, but you can keep entering digits.

Just throwing around ideas here.

The PIN idea is great as most security sensitive apps include the feature.

**Comment by storyjesse** [comment]
> Basic PIN for all apps, and a setting to set how many minutes until it is set. While PIN-locked, terminal keybase commands are also restricted.
> The PIN idea is great as most security sensitive apps include the feature.

Since keybase is used on phones and laptops a PIN is absolutely needed. Imagine if you were on a busy train and your phone was snatched out of your hand just as the doors were closing. Someone now has access to all your chat conversations and history.
Or imagine you are an activist trying to stop a coal mine from being built and your phone is confiscated by police. If they manage to unlock your phone they will now have access to all your keybase chat history, which could be detrimental for people besides yourself.
A simple PIN lock with adjustable timeout is VITAL for keybase to be a serious secure messaging tool.

**Comment by networkException** [comment]
are there any updates or plans on implementing this in a future release? Perhaps a biometric lock for the app?
